### DISCUSSION TITLE
VEAS not using ISA Std Day sea-level density?

### DISCUSSION THREAD

**Discussion by seanmcleod** [JSBSim-Team/jsbsim#894] [General]
While reviewing PR - https://github.com/JSBSim-Team/jsbsim/pull/885 I noticed that the calculation for `VEAS` doesn't appear to be using the ISA Std Day sea-level density in it's calculation?

https://github.com/JSBSim-Team/jsbsim/blob/92fa1eaa29322dc3f6ab55587161e55a6b24d36d/src/models/FGAuxiliary.cpp#L201

Double-checking `in.DensitySL` it isn't the standard day value, except on a standard day.

See EAS formulas - https://en.wikipedia.org/wiki/Equivalent_airspeed

**Comment by gallonmate** [comment]
> it isn't the standard day value, except on a standard day.

This made me chuckle even if it was unintended pun. :laughing: 

Looks like you found an issue. But I'm curious which methods you use to setup a non-standard day? 

Like you said by default JSBSim uses standard day values and everything is fine: 
FGAuxiliary`DensitySL` = FGAtmosphere`SLDensity` = FGStandardAtmosphere `StdSLdensity`
FGStandardAtmosphere provides correct sea level density:
https://github.com/JSBSim-Team/jsbsim/blob/92fa1eaa29322dc3f6ab55587161e55a6b24d36d/src/models/atmosphere/FGStandardAtmosphere.cpp#L133-L134

The issue I assume you mean and the issue I see is when using SetTemperature of FGStandardAtmosphere to change conditions at a flight location. This ends up adding a temperature bias to SLTemperature and then recalculation of Density is called which uses the new offset SLTemperature.

https://github.com/JSBSim-Team/jsbsim/blob/ebb770ebc0ec4cee6aefef6ad88702ef1164b5d7/src/models/atmosphere/FGStandardAtmosphere.cpp#L328

```c++
void FGStandardAtmosphere::SetTemperature(double t, double h, eTemperature unit)
{
  double targetTemp = ConvertToRankine(t, unit);
  double GeoPotAlt = GeopotentialAltitude(h);
  double bias = targetTemp - GetStdTemperature(h);

  if (GeoPotAlt <= GradientFadeoutAltitude)
    bias -= TemperatureDeltaGradient * (GradientFadeoutAltitude - GeoPotAlt);

  SetTemperatureBias(eRankine, bias);
  CalculatePressureBreakpoints(SLpressure);

  SLtemperature = GetTemperature(0.0);
  CalculateSLSoundSpeedAndDensity();
}
```

Is there a proper way of defining non-standard day conditions? I'm not sure if there are multiple issues or just the one you found.
 
The obvious solution is to hardcode density like you alluded to `veas = sqrt(2 * qbar / 0.0023768924)`, but in doing so, does this limit JSBSim in some way, perhaps on non-earth planets? Is EAS even relevant to other planets :alien: :flying_saucer: ?

And if Set Temperature is a correct way of defining local conditions for flying, then should it even be recalculating sea level density at all? Unless this has no affect other than VEAS, which can be fixed easily, then I guess Set Temperature does not need correction, only VEAS does.

**Reply by bcoconni** [reply]
> but in doing so, does this limit JSBSim in some way, perhaps on non-earth planets? Is EAS even relevant to other planets :alien: :flying_saucer: ?

JSBSim has had the ambition for a long time to be an interplanetary FDM [as is shown by the existence of the Mars atmosphere model in our codebase since 2004](https://github.com/JSBSim-Team/jsbsim/blob/master/src/models/atmosphere/FGMars.cpp) :smile: Unfortunately this code is inoperative at the moment as it is currently not possible to modify the atmosphere model at run time but this is in the process of being changed (see the PR #902 and #908).

**Comment by seanmcleod** [comment]
> This made me chuckle even if it was unintended pun.

It was intended, good to know it made someone chuckle 😉 

Yes, the ways I'm familiar with in terms of setting non-standard conditions is to make use of the temperature setting options:

```c++
  PropertyManager->Tie("atmosphere/delta-T", this, eRankine,
                                    (PMFi)&FGStandardAtmosphere::GetTemperatureBias,
                                    (PMF)&FGStandardAtmosphere::SetTemperatureBias);
  PropertyManager->Tie("atmosphere/SL-graded-delta-T", this, eRankine,
                                    (PMFi)&FGStandardAtmosphere::GetTemperatureDeltaGradient,
                                    (PMF)&FGStandardAtmosphere::SetSLTemperatureGradedDelta);
```

And I guess:

```c++
  PropertyManager->Tie("atmosphere/P-sl-psf", this, ePSF,
                                   (PMFi)&FGStandardAtmosphere::GetPressureSL,
                                   (PMF)&FGStandardAtmosphere::SetPressureSL);
```

> Is EAS even relevant to other planets

Yep, IAS/EAS is very much tied based on calibration etc. to specific earth conditions. Although I guess if a pilot knows the stall speed is 100KIAS on earth they can use the same airspeed indicator while flying the same aircraft on Mars and know they can use the 100KIAS reference for stall?

**Comment by gallonmate** [comment]
It appears all those Set Temp, and Set Pressure functions end up updating SLDensity, affecting VEAS. Before this 2018 update, https://github.com/JSBSim-Team/jsbsim/commit/3e53a34fd665e4b131deb74bec6101f12c4841bc  Density was not calculated when using set tempertuare/pressure in FGStandardAtmosphere. 

I guess that leaves the question would creating some new Set Temp/ Set Pressure functions without affecting sea level density be useful or would it just be better to update VEAS to use a new variable that is a "Standard" sea level density, separate from SLDensity.

**Reply by bcoconni** [reply]
> It appears all those Set Temp, and Set Pressure functions end up updating SLDensity, affecting VEAS. Before this 2018 update, [3e53a34](https://github.com/JSBSim-Team/jsbsim/commit/3e53a34fd665e4b131deb74bec6101f12c4841bc) Density was not calculated when using set tempertuare/pressure in FGStandardAtmosphere.
> 
> I guess that leaves the question would creating some new Set Temp/ Set Pressure functions without affecting sea level density be useful or would it just be better to update VEAS to use a new variable that is a "Standard" sea level density, separate from SLDensity.

Well, from my point of view pressure, temperature and density should be consistent at any time according to the ideal gas law $P=\rho RT$ and that was the very goal of the commit you mentioned. So as @seanmcleod suggested below, the methods `SetTemperature()` and `SetPressure()` should be left as they currently are.

**Comment by seanmcleod** [comment]
I would suggest leaving the `SetTemp()`, `SetPressure()` functions as is, and expose and use `StdDaySLdensity` just like we use `StdDaySLpressure` and `StdDaySLsoundspeed` in converting between Mach and CAS.

https://github.com/JSBSim-Team/jsbsim/blob/92fa1eaa29322dc3f6ab55587161e55a6b24d36d/src/models/FGAtmosphere.cpp#L367-L381

For interest there is an equivalent calculation of EAS making using of `StdDaySLpressure` and `StdDaySLsoundspeed`.

$$ EAS = a_0 M \sqrt{\frac{P}{P_0}} $$

**Reply by bcoconni** [reply]
> I would suggest leaving the `SetTemp()`, `SetPressure()` functions as is, and expose and use `StdDaySLdensity` just like we use `StdDaySLpressure` and `StdDaySLsoundspeed` in converting between Mach and CAS.
> 
> https://github.com/JSBSim-Team/jsbsim/blob/92fa1eaa29322dc3f6ab55587161e55a6b24d36d/src/models/FGAtmosphere.cpp#L367-L381

I second that 👍
Would you mind submitting a PR to fix this issue ?

**Comment by seanmcleod** [comment]
> Would you mind submitting a PR to fix this issue ?

Will do, although my plan was to wait until PR https://github.com/JSBSim-Team/jsbsim/pull/885 was resolved.

**Comment by bcoconni** [comment]
FYI I have submitted the PR #909 to fix the very same issue in `FGInitialCondition`.
